pkg/cloudprovider/providers/vsphere/vclib/datastore.go

```diff
@@ -0,0 +1,140 @@
+			if _, ok := soapFault.VimFault().(types.FileAlreadyExists); ok {
+				return ErrFileAlreadyExist
+			}
+		}
+		return err
+	}
+	glog.V(LOG_LEVEL).Infof("Created dir with path as %+q", directoryPath)
+	return nil
+}
+
+// Create a virtual disk at given diskPath using specified values in the volumeOptions object
+func (ds DataStore) CreateVirtualDisk(ctx context.Context, diskPath string, volumeOptions VolumeOptions) (err error) {
+	if !volumeOptions.VerifyVolumeOptions() {
+		glog.Error("VolumeOptions verification failed")
+		return ErrInvalidVolumeOptions
+	}
+	// Create virtual disk
+	diskFormat := diskFormatValidType[volumeOptions.DiskFormat]
+	// Create a virtual disk manager
+	virtualDiskManager := object.NewVirtualDiskManager(ds.Client())
+	// Create specification for new virtual disk
+	vmDiskSpec := &types.FileBackedVirtualDiskSpec{
+		VirtualDiskSpec: types.VirtualDiskSpec{
+			AdapterType: LSILogicControllerType,
+			DiskType:    diskFormat,
+		},
+		CapacityKb: int64(volumeOptions.CapacityKB),
+	}
+	task, err := virtualDiskManager.CreateVirtualDisk(ctx, diskPath, nil, vmDiskSpec)
+	if err != nil {
+		glog.Errorf("Failed to create virtual disk: %s. err %v", diskPath, err)
+		return err
+	}
+	return task.Wait(ctx)
+}
+
+// Creates a virtual disk with the policy configured to the disk.
+// A call to this function is made only when a user specifies VSAN storage capabilties in the storage class definition.
+func (ds DataStore) CreateVirtualDiskWithPolicy(ctx context.Context, diskPath string, virtualMachine VirtualMachine, diskControllerType string, volumeOptions VolumeOptions) error {
+
+	disk, _, err := CreateDiskSpec(ctx, virtualMachine, ds.Reference(), diskPath, diskControllerType, volumeOptions)
+	if err != nil {
+		glog.Errorf("Failed to create Disk Spec. err: %v", err)
+		return err
+	}
+	// Reconfigure VM
+	virtualMachineConfigSpec := types.VirtualMachineConfigSpec{}
+	deviceConfigSpec := &types.VirtualDeviceConfigSpec{
+		Device:        disk,
+		Operation:     types.VirtualDeviceConfigSpecOperationAdd,
+		FileOperation: types.VirtualDeviceConfigSpecFileOperationCreate,
+	}
+	storageProfileSpec := &types.VirtualMachineDefinedProfileSpec{}
+	// Is PBM storage policy ID is present, set the storage spec profile ID,
+	// else, set raw the VSAN policy string.
+	if volumeOptions.StoragePolicyID != "" {
+		storageProfileSpec.ProfileId = volumeOptions.StoragePolicyID
+	} else if volumeOptions.VSANStorageProfileData != "" {
+		storageProfileSpec.ProfileId = ""
+		storageProfileSpec.ProfileData = &types.VirtualMachineProfileRawData{
+			ExtensionKey: "com.vmware.vim.sps",
+			ObjectData:   volumeOptions.VSANStorageProfileData,
+		}
+	}
+	deviceConfigSpec.Profile = append(deviceConfigSpec.Profile, storageProfileSpec)
+	virtualMachineConfigSpec.DeviceChange = append(virtualMachineConfigSpec.DeviceChange, deviceConfigSpec)
+	task, err := virtualMachine.Reconfigure(ctx, virtualMachineConfigSpec)
+	if err != nil {
```

Comment:
Two questions.
1. This looks has some common code with VirtualMachine.AttachDisk() function. not sure if you want a common function here.
2. If not, then the VirtualMachine.AttachDisk() function called vm.DeleteController when vm.Reconfigure() returns with error. But this step is not included here. Not sure if it's not necessary in this case.